core/tests/test-dependencies/test-to-modules-matcher.ts

```diff
@@ -0,0 +1,228 @@
+      'core/tests/test-modules-mappings/acceptance/logged-in-user/manage-goals-progress-and-lessons-from-learner-dashboard.txt',
+    ],
+};
+
+const LOCALHOST_URL = 'http://localhost:8181/';
+
+const matchUrl = (url: string, route: Route): boolean => {
+  if (route.path === url) {
+    return true;
+  }
+
+  const urlSerializer = new DefaultUrlSerializer();
+  const urlTree = urlSerializer.parse(url);
+  if (!urlTree.root.children.primary) {
+    return false;
+  }
+
+  const segments: UrlSegment[] = urlTree.root.children.primary.segments;
+  const segmentGroup: UrlSegmentGroup = urlTree.root.children.primary;
+
+  if (route.path === undefined) {
+    return false;
+  }
+
+  const parts = route.path.split('/');
+  if (parts.length > segments.length) {
+    return false;
+  }
+
+  if (
+    route.pathMatch === 'full' &&
+    (segmentGroup.hasChildren() || parts.length < segments.length)
+  ) {
+    return false;
+  }
+
+  for (let index = 0; index < parts.length; index++) {
+    const part = parts[index];
+    const segment = segments[index];
+    const partIsParameter = part.startsWith(':');
+    if (partIsParameter) {
+      continue;
+    } else if (part !== segment.path) {
+      return false;
+    }
+  }
+
+  return true;
+};
+
+export class TestToModulesMatcher {
+  static goldenFilePath: string;
+  static routeToModuleMapping: Map<Route, string> = getRouteToModuleMapping();
+  static collectedModules: string[] = [];
+  static collectedErrors: string[] = [];
+
+  /**
+   * Sets the path of the golden file that contains the list of Angular modules
+   * that are expected to be used by tests.
+   */
+  public static setGoldenFilePath(filePath: string): void {
+    if (!filePath.endsWith('.txt')) {
+      throw new Error('The golden file path must have a .txt extension.');
+    }
+    this.goldenFilePath = filePath;
+  }
+
+  /**
+   * Registers a URL and matches it with an Angular module.
+   */
+  public static registerUrl(url: string): void {
+    if (!url.startsWith(LOCALHOST_URL)) {
+      return;
+    }
+    const path = url.replace(LOCALHOST_URL, '');
+    let matched = false;
+    for (const [route, module] of this.routeToModuleMapping.entries()) {
+      if (!matchUrl(path, route)) {
+        continue;
+      }
+      matched = true;
+      if (this.collectedModules.includes(module)) {
+        break;
+      }
+      if (COMMON_MODULES_TO_EXCLUDE[module]) {
+        let exclude = true;
+        for (const globPattern of COMMON_MODULES_TO_EXCLUDE[module]) {
+          const globMatcher = new glob.GlobSync(globPattern);
+          exclude = !globMatcher.minimatch.match(this.goldenFilePath);
+          if (!exclude) {
+            break;
+          }
+        }
+
+        if (exclude) {
+          break;
+        }
+      }
+      this.collectedModules.push(module);
+      break;
+    }
+
+    if (!matched) {
+      const errorMessage = `No Angular module found for URL: ${url}`;
+      if (!this.collectedErrors.includes(errorMessage)) {
```

Comment:
Should we add more info to this error message?

Reply:
Done.